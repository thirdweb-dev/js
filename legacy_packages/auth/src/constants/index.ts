const THIRDWEB_AUTH_COOKIE_PREFIX = `thirdweb_auth`;
export const THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX = `${THIRDWEB_AUTH_COOKIE_PREFIX}_token`;
export const THIRDWEB_AUTH_ACTIVE_ACCOUNT_COOKIE = `${THIRDWEB_AUTH_COOKIE_PREFIX}_active_account`;

export const THIRDWEB_AUTH_DEFAULT_LOGIN_PAYLOAD_DURATION_IN_SECONDS = 60 * 10;
export const THIRDWEB_AUTH_DEFAULT_TOKEN_DURATION_IN_SECONDS = 60 * 60 * 24;
export const THIRDWEB_AUTH_DEFAULT_REFRESH_INTERVAL_IN_SECONDS = 5;
