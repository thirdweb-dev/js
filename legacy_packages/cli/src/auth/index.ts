import assert from "assert";
import chalk from "chalk";
import fs from "fs";
import http from "http";
import open from "open";
import ora from "ora";
import url from "url";
import { logger, spinner } from "../core/helpers/logger";
import { ThirdwebAuth } from "@thirdweb-dev/auth";
import { ICredsConfig } from "../lib/types";
import { generateStateParameter } from "../lib/utils";
import crypto from "node:crypto";

type LoginProps = {
  browser: boolean;
  configPaths: ConfigPaths;
};

const defaultLoginProps = {
  browser: true,
  configPaths: {
    credsConfigPath: "",
    cliWalletPath: "",
    tokenPath: "",
  },
};

type ConfigPaths = {
  credsConfigPath: string;
  cliWalletPath: string;
  tokenPath: string;
};

export async function loginUser(
  configPaths: ConfigPaths,
  options?: { new: boolean },
  showLogs?: boolean,
) {
  const { credsConfigPath, tokenPath } = configPaths;
  const authToken = await getSession(tokenPath, credsConfigPath);
  if (authToken && !options?.new) {
    if (showLogs) {
      console.log(chalk.green("You are already logged in"));
    }
    globalThis["TW_CLI_AUTH_TOKEN"] = authToken;
    return authToken;
  } else {
    const token = await authenticateUser({ browser: true, configPaths });
    if (!token) {
      throw new Error("Failed to login");
    }

    globalThis["TW_CLI_AUTH_TOKEN"] = token;
    return token;
  }
}

export async function logoutUser(
  credsConfigPath: string,
  tokenPath: string,
  cliWalletPath: string,
) {
  try {
    ora("Logging out...").start();
    const dirExists =
      fs.existsSync(credsConfigPath) &&
      fs.existsSync(tokenPath) &&
      fs.existsSync(cliWalletPath);
    if (!dirExists) {
      ora().warn(
        chalk.yellow("You are already logged out, did you mean to login?"),
      );
      return;
    }
    fs.unlinkSync(credsConfigPath);
    fs.unlinkSync(tokenPath);
    // TODO: We can consider not removing this on logout later, once we want to implement teams. For now this wallet will be ephemeral.
    fs.unlinkSync(cliWalletPath);
    ora().succeed(chalk.green("You have been logged out"));
  } catch (error) {
    console.log(chalk.red("Something went wrong", error));
  }
}

export async function getSession(tokenPath: string, configCredsPath: string) {
  if (!fs.existsSync(tokenPath) || !fs.existsSync(configCredsPath)) {
    return null;
  }
  try {
    return fs.readFileSync(tokenPath, "utf8");
  } catch (error) {
    console.error(error);
  }
}

export const authenticateUser = async (
  props: LoginProps = defaultLoginProps,
) => {
  const { credsConfigPath, cliWalletPath, tokenPath } = props.configPaths;
  const waitForDashboard = spinner(
    "Waiting for a response from the dashboard",
  ).clear();

  // Get or generate a localwallet.
  const wallet = await getOrGenerateLocalWallet(credsConfigPath, cliWalletPath);
  const walletAddress = await wallet.getAddress();
  const domain = "thirdweb.com";
  const auth = new ThirdwebAuth(wallet, domain);

  // Generate the login payload to pass to the dashboard.
  const loggedIn = await auth.login({
    domain: domain,
    address: walletAddress,
  });

  // In this case the state is the loggedIn object.
  const ourState = generateStateParameter(32);
  const payload = encodeURIComponent(JSON.stringify(loggedIn));
  const urlToOpen = `https://thirdweb.com/cli/login?payload=${payload}&#${ourState}`;

  let server: http.Server;
  let loginTimeoutHandle: NodeJS.Timeout;
  const timerPromise = new Promise<void>((resolve, reject) => {
    loginTimeoutHandle = setTimeout(() => {
      logger.error(
        "Login session timed out, server didn't receive a response in 5 minutes. Please try again.",
      );
      server.close();
      clearTimeout(loginTimeoutHandle);
      reject(
        new Error(
          "Login session timed out, server didn't receive a response in 5 minutes. Please try again.",
        ),
      );
    }, 300000);
  });

  const loginPromise = new Promise<string>((resolve, reject) => {
    server = http.createServer(async (req, res) => {
      function finish(error?: Error) {
        clearTimeout(loginTimeoutHandle);
        server.close((closeErr?: Error) => {
          if (error || closeErr) {
            reject(error || closeErr);
          }
        });
      }
      res.setHeader("Access-Control-Allow-Credentials", "true");
      res.setHeader("Access-Control-Allow-Origin", "https://thirdweb.com");
      res.setHeader("Access-Control-Allow-Methods", "POST");

      if (req.method === "OPTIONS") {
        res.setHeader(
          "Access-Control-Allow-Headers",
          "content-type, baggage, sentry-trace",
        );
        res.writeHead(200);
        res.end();
        return;
      }

      assert(req.url, "This request doesn't have a URL");
      const { pathname, query } = url.parse(req.url, true);
      switch (pathname) {
        case "/auth/callback": {
          if (query.failed) {
            res.writeHead(500, "Unable to authenticate with the dashboard!");
            res.end("Unable to authenticate with the dashboard!", () => {
              finish(new Error("Unable to authenticate with the dashboard!"));
            });
            reject(
              chalk.red(
                "Something went wrong! Unable to authenticate with the dashboard.",
              ),
            );
            waitForDashboard.stop();
          }
          if (query.token) {
            const token = Array.isArray(query.token)
              ? query.token[0]
              : query.token;
            const theirState = Array.isArray(query.state)
              ? query.state[0]
              : query.state;

            // Check if the state sent back is the same as the one we sent.
            if (theirState !== ourState) {
              res.writeHead(400, { "Content-Type": "text/plain" }); // send a 400 response
              res.end("Unauthorized request, state mismatch", () => {
                finish(new Error("Unauthorized request, state mismatch"));
              });
              reject(
                new Error(chalk.red("\nUnauthorized request, state mismatch")),
              );
              waitForDashboard.stop();
            } else {
              // Save the token to the config file.
              // eslint-disable-next-line no-unused-expressions
              fs.writeFileSync(tokenPath, token),
                {
                  encoding: "utf8",
                  mode: 0o600,
                };
              res.end(() => {
                waitForDashboard.clear();
                console.log(
                  chalk.green(
                    `Successfully linked your account to this device`,
                  ),
                );
                finish();
              });
              logger.info(chalk.green(`\nSuccessfully logged in.`));
              resolve(token); // resolve promise with secretKey
            }
          } else {
            res.writeHead(400, { "Content-Type": "text/plain" });
            res.end("No authToken received", () => {
              finish(new Error("No authToken received"));
            });
            reject(new Error(chalk.red("\nNo authToken received")));
            waitForDashboard.stop();
          }
        }
      }
    });

    server.listen(8976);
  });
  console.log(
    `Automatically attempting to open a link to authenticate with our dashboard...\n`,
  );
  waitForDashboard.start();
  // Adding this timeout since it feels weird for the browser to open before the spinner.
  setTimeout(async () => {
    await open(urlToOpen);
  }, 2000);

  console.log(
    chalk.yellow(
      `If the browser doesn't open, please use this link to authenticate:\n`,
    ),
  );
  console.log(chalk.yellow(urlToOpen + "\n"));

  return Promise.race([timerPromise, loginPromise]);
};

async function getOrCreatePassword(configCredsPath: string): Promise<string> {
  const passwordFileExists = fs.existsSync(configCredsPath);
  if (!passwordFileExists) {
    // Generate a random password for the CLI to be able to import / export the wallet.
    return crypto.randomUUID();
  } else {
    const file = fs.readFileSync(configCredsPath, "utf8");
    try {
      JSON.parse(file);
    } catch (e) {
      // If it fails to parse the password we should just delete the file and generate a new password, since this password is not shown to the user.
      fs.unlinkSync(configCredsPath);
      return getOrCreatePassword(configCredsPath);
    }
    const { password } = JSON.parse(file) as ICredsConfig;
    return password;
  }
}

async function getOrGenerateLocalWallet(
  configCredsPath: string,
  cliWalletPath: string,
) {
  // Get or prompt for password.
  const password = await getOrCreatePassword(configCredsPath);
  const { LocalWallet } = await import("@thirdweb-dev/wallets");
  const wallet = new LocalWallet();
  const foundWallet = fs.existsSync(cliWalletPath);

  if (foundWallet) {
    const walletJson = fs.readFileSync(cliWalletPath, "utf8");
    try {
      // See if file is valid before proceeding.
      JSON.parse(walletJson);

      await wallet.import({
        encryptedJson: walletJson,
        password,
      });
      return wallet;
    } catch (e) {
      // Wallet file is not valid json, create a new one.
    }
  }

  // Otherwise, generate a new wallet.
  wallet.generate();
  const walletExported = await wallet.export({
    strategy: "encryptedJson",
    password,
  });

  // write wallet
  fs.writeFileSync(cliWalletPath, walletExported, {
    encoding: "utf8",
    mode: 0o600,
  });

  // write password
  fs.writeFileSync(
    configCredsPath,
    JSON.stringify({
      password: password,
    }),
    "utf8",
  );

  return wallet;
}

export const validateKey = async (apiSecretKey: string) => {
  const apiUrl = "https://api.thirdweb.com/v1/keys/use";
  try {
    const response = await fetch(`${apiUrl}?scope=storage`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        "x-secret-key": apiSecretKey,
      },
    });

    if (response.status !== 200) {
      throw new Error("Unauthorized key");
    }
  } catch (error) {
    throw new Error(chalk.red("Unauthorized key"));
  }
};
