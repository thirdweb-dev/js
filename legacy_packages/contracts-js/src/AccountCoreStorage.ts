/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface AccountCoreStorageInterface extends utils.Interface {
  functions: {
    "ACCOUNT_CORE_STORAGE_POSITION()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "ACCOUNT_CORE_STORAGE_POSITION"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "ACCOUNT_CORE_STORAGE_POSITION",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "ACCOUNT_CORE_STORAGE_POSITION",
    data: BytesLike
  ): Result;

  events: {};
}

export interface AccountCoreStorage extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: AccountCoreStorageInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    ACCOUNT_CORE_STORAGE_POSITION(overrides?: CallOverrides): Promise<[string]>;
  };

  ACCOUNT_CORE_STORAGE_POSITION(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    ACCOUNT_CORE_STORAGE_POSITION(overrides?: CallOverrides): Promise<string>;
  };

  filters: {};

  estimateGas: {
    ACCOUNT_CORE_STORAGE_POSITION(
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    ACCOUNT_CORE_STORAGE_POSITION(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
