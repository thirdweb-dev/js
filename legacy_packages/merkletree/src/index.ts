export { MerkleTree } from "./merkletree/MerkleTree";
