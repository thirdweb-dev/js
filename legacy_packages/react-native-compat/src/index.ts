// react-native-get-random-values needs to be first imported module
// prettier-ignore
import "react-native-get-random-values";

// URL polyfill for React Native
import 'react-native-url-polyfill/auto';

import "./shims";
