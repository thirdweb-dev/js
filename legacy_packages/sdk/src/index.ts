// handle browser vs node global

// eslint-disable-next-line better-tree-shaking/no-top-level-side-effects
globalThis.global = globalThis;

export type {
  NFTMetadataInput,
  NFT,
  NFTMetadata,
  NFTMetadataOrUri,
  BasicNFTInput,
} from "./core/schema/nft";

export type { CurrencyValue, TokenMetadata } from "./core/schema/token";
export { getRpcUrl } from "./core/constants/urls";

export { checkClientIdOrSecretKey } from "./core/utils/apiKey";

export type { QueryAllParams } from "./core/schema/QueryParams";

// export EVM by default
export * from "./evm";
