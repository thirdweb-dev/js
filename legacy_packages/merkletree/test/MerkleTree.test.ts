import { MerkleTree } from "../src/index";
import { keccak256Sync, sha256Sync } from "@thirdweb-dev/crypto";

describe("MerkleTree", () => {
  it("sha256 with keccak256 leaves", () => {
    const leaves = ["a", "b", "c"].map(keccak256Sync);
    const tree = new MerkleTree(leaves, sha256Sync);

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
    // check equality of the leaves
    expect(tree.getHexLeaves()).toMatchSnapshot();
    // check equality of the layers
    expect(tree.getHexLayers()).toMatchSnapshot();
  });

  it("sha256 with keccak256 leaves with duplicate odd option", () => {
    const leaves = ["a", "b", "c"].map(keccak256Sync);
    const tree = new MerkleTree(leaves, sha256Sync, { duplicateOdd: true });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
    // check equality of the leaves
    expect(tree.getHexLeaves()).toMatchSnapshot();
    // check equality of the layers
    expect(tree.getHexLayers()).toMatchSnapshot();
  });

  it("sha256 with sort pairs option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map(sha256Sync);
    const tree = new MerkleTree(leaves, sha256Sync, { sortPairs: true });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  it("sha256 verify with positional hex proof and no pairSort", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map(sha256Sync);
    const tree = new MerkleTree(leaves, sha256Sync, { sortPairs: false });

    expect(
      tree.verify(
        // TODO: accept Uint8Array as input
        tree.getProof(Buffer.from(leaves[1]), 1),
        // TODO: accept Uint8Array as input
        Buffer.from(leaves[1]),
        tree.getHexRoot(),
      ),
    ).toBe(true);
  });

  test("sha256 verify with non-hex proof and no pairSort", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map(sha256Sync);
    const tree = new MerkleTree(leaves, sha256Sync, { sortPairs: false });

    expect(
      tree.verify(
        // TODO: accept Uint8Array as input
        tree.getProof(Buffer.from(leaves[1]), 1),
        // TODO: accept Uint8Array as input
        Buffer.from(leaves[1]),
        tree.getHexRoot(),
      ),
    ).toBe(true);
  });

  it("sha256 verify with hex proof and pairSort", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map(sha256Sync);
    const tree = new MerkleTree(leaves, sha256Sync, { sortPairs: true });

    expect(
      tree.verify(
        // TODO: accept Uint8Array as input
        tree.getHexProof(Buffer.from(leaves[1]), 1),
        // TODO: accept Uint8Array as input
        Buffer.from(leaves[1]),
        tree.getHexRoot(),
      ),
    ).toBe(true);
  });

  it("keccak256 with sort leaves and sort pairs option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"].map(
      keccak256Sync,
    );
    const tree = new MerkleTree(leaves, keccak256Sync, {
      sortLeaves: true,
      sortPairs: true,
    });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  it("keccak256 with sort option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"].map(
      keccak256Sync,
    );
    const tree = new MerkleTree(leaves, keccak256Sync, { sort: true });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  it("sha256 with sha256 leaves and sort pairs option and duplicate odd option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map(sha256Sync);
    const tree = new MerkleTree(leaves, sha256Sync, {
      sortPairs: true,
      duplicateOdd: true,
    });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  it("sha256 with hash leaves option", () => {
    const expectedRoot =
      "0x1f7379539707bcaea00564168d1d4d626b09b73f8a2a365234c62d763f854da2";

    const leaves = ["a", "b", "c", "d", "e", "f"];
    const tree1 = new MerkleTree(leaves, sha256Sync, { hashLeaves: true });
    expect(tree1.getHexRoot()).toEqual(expectedRoot);

    const hashedLeaves = ["a", "b", "c", "d", "e", "f"].map(sha256Sync);
    const tree2 = new MerkleTree(hashedLeaves, sha256Sync);
    expect(tree2.getHexRoot()).toEqual(expectedRoot);
  });

  it("sha256 with hash leaves option and duplicate odd option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"];
    const tree = new MerkleTree(leaves, sha256Sync, {
      hashLeaves: true,
      duplicateOdd: true,
    });
    const root =
      "44205acec5156114821f1f71d87c72e0de395633cd1589def6d4444cc79f8103";

    expect(tree.getRoot().toString("hex")).toEqual(root);
  });

  it("solidity keccak256", () => {
    const leaves = ["a", "b", "c"].map(keccak256Sync);

    const a_hash =
      "3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb";
    const b_hash =
      "b5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510";
    const c_hash =
      "0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2";

    expect(leaves.map((l) => Buffer.from(l).toString("hex"))).toEqual([
      a_hash,
      b_hash,
      c_hash,
    ]);

    const tree = new MerkleTree(leaves, keccak256Sync);

    const layers = tree.getLayers().slice(1); // no leaves

    const layer_1 = Buffer.from(
      keccak256Sync(Buffer.concat([leaves[0], leaves[1]])),
    ).toString("hex");

    expect(layers[0][0].toString("hex")).toEqual(layer_1);
    expect(layers[0][1].toString("hex")).toEqual(c_hash);

    const root = Buffer.from(
      "aff1208e69c9e8be9b584b07ebac4e48a1ee9d15ce3afe20b77a4d29e4175aa3",
      "hex",
    );
    expect(tree.getRoot().toString("hex")).toEqual(root.toString("hex"));

    const proof_0 = tree.getProof(Buffer.from(leaves[0]));
    expect(proof_0.length).toEqual(2);
    expect(proof_0[0].position).toEqual("right");
    expect(proof_0[0].data.toString("hex")).toEqual(b_hash);
    expect(proof_0[1].position).toEqual("right");
    expect(proof_0[1].data.toString("hex")).toEqual(c_hash);

    expect(tree.verify(proof_0, Buffer.from(leaves[0]), root)).toBe(true);

    const proof_1 = tree.getProof(Buffer.from(leaves[1]));
    expect(proof_1.length).toEqual(2);
    expect(proof_1[0].position).toEqual("left");
    expect(proof_1[0].data.toString("hex")).toEqual(a_hash);
    expect(proof_1[1].position).toEqual("right");
    expect(proof_1[1].data.toString("hex")).toEqual(c_hash);

    expect(tree.verify(proof_1, Buffer.from(leaves[1]), root)).toBe(true);

    const proof_2 = tree.getProof(Buffer.from(leaves[2]));
    expect(proof_2.length).toEqual(1);
    expect(proof_2[0].position).toEqual("left");
    expect(proof_2[0].data.toString("hex")).toEqual(layer_1);

    expect(tree.verify(proof_2, Buffer.from(leaves[2]), root)).toBe(true);
  });
});
