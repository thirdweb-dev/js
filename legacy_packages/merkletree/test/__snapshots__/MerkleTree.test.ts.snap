// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`MerkleTree sha256 with keccak256 leaves 1`] = `"0x311d2e46f49b15fff8b746b74ad57f2cc9e0d9939fda94387141a2d3fdf187ae"`;

exports[`MerkleTree sha256 with keccak256 leaves 2`] = `
[
  "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
  "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
  "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
]
`;

exports[`MerkleTree sha256 with keccak256 leaves 3`] = `
[
  [
    "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
    "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
    "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
  ],
  [
    "0x176f0f307632fdd5831875eb709e2f68d770b102262998b214ddeb3f04164ae1",
    "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
  ],
  [
    "0x311d2e46f49b15fff8b746b74ad57f2cc9e0d9939fda94387141a2d3fdf187ae",
  ],
]
`;

exports[`MerkleTree sha256 with keccak256 leaves with duplicate odd option 1`] = `"0xbcdd0f60308db788712205115fe4273bfda49fa0925611fee765a63df9ab96a1"`;

exports[`MerkleTree sha256 with keccak256 leaves with duplicate odd option 2`] = `
[
  "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
  "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
  "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
]
`;

exports[`MerkleTree sha256 with keccak256 leaves with duplicate odd option 3`] = `
[
  [
    "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
    "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
    "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
  ],
  [
    "0x176f0f307632fdd5831875eb709e2f68d770b102262998b214ddeb3f04164ae1",
    "0x43e061172b1177f25d0f156b2d2ed11728006fade8e167ff3d1b9dbc979a3358",
  ],
  [
    "0xbcdd0f60308db788712205115fe4273bfda49fa0925611fee765a63df9ab96a1",
  ],
]
`;

exports[`MerkleTree sha256 with sort pairs option 1`] = `"0xa30ba95a1a5dc397fe45ea20105363b08d682b864a28f4940419a29349a28325"`;

exports[`MerkleTree keccak256 with sort option 1`] = `"0x60219f87561939610b484575e45c6e81156a53b86d7cd16640d930d14f21758e"`;

exports[`MerkleTree sha256 with sha256 leaves and sort pairs option and duplicate odd option 1`] = `"0xa5260b2a7ec31584e5d5689a5628c2b3d949e2397334fd71c107478e5f887eaf"`;

exports[`MerkleTree keccak256 with sort leaves and sort pairs option 1`] = `"0x60219f87561939610b484575e45c6e81156a53b86d7cd16640d930d14f21758e"`;
